### PR TITLE
Initial commit of piet-wgsl

### PR DESCRIPTION
**Pull request by raphlinus**
Starting an experimental port to WGSL shader language, using wgpu to run the examples. As of this commit, it's quite hacky and takes some shortcuts, but does render paths to a grayscale texture.

### CODE REVIEW

**Review by DJMcNab — APPROVED** [review]
I love that there's no `unsafe` required for this.
It also works on my machine with the small change to make the template directory independent of the running method.

One issue I have with working on this is the other crates don't compile because of the missing, however removing them from the workspace temporarily works for me.

The other challenge is IDE support for the shaders. Ideally, we'd be able to use [wgsl-analyzer](https://marketplace.visualstudio.com/items?itemName=wgsl-analyzer.wgsl-analyzer), but that doesn't like this templating format. If you were willing, I'd be happy to migrate to the form it understands; probably by extracting the important code from bevy, depending on how wgsl-analyzer does it.

The other option is to get wgsl-analyzer to understand handlebars syntax

File: piet-wgsl/shader/backdrop.twgsl [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Also licensed under MIT license, at your choice.
+
+// Note: this is the non-atomic version
+struct Tile {
```
[the remark this answers is not in the document]
Comment (DJMcNab) [reply piet-wgsl/shader/backdrop.twgsl:18]: Ah, this is https://github.com/linebender/piet-gpu/pull/105 having an unintended consequence.

File: piet-wgsl/README.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# piet-wgsl
+
+This crate is currently a highly experimental proof-of-concept port of the piet-gpu renderer to the WGSL shader language, so it could be run on WebGPU. Depending on how well it works out, it may become the authoritative source for piet-gpu.
+
+The shaders are actually handlebars templates over WGSL, as it's important to share common data structures; it's likely we'll use the template mechanism to supply various parameters which are not supported by the WGSL language, for example to specify grayscale or RGBA buffer output for fine rasterization.
```
[the remark this answers is not in the document]
Comment (DJMcNab) [reply piet-wgsl/README.md:5]: > I don't know, maybe it's feasible if someone were to put in a lot of effort, but its not easy unfortunately. The problem is that if you just naively preprocess the files before handing them over to wgsl-analyzer, then all the offsets and spans are wrong, so error messages are all over the place. This is kind of handled for #imports, but in a kind of hacky way and it would be much harder to do the same for general handlebars support, not just {{> config}}.
> So short answer, not really that feasable

@jakobhellermann on adding handlebars support to wgsl-analyzer.

File: piet-wgsl/shader/backdrop.twgsl [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+@group(0) @binding(1)
+var<storage, read_write> tiles: array<Tile>;
+
+let WG_SIZE = 64u;
+
+var<workgroup> sh_backdrop: array<i32, WG_SIZE>;
+
+// Each workgroup computes the inclusive prefix sum of the backdrops
+// in one row of tiles.
```
[the remark this answers is not in the document]
Comment (raphlinus) [reply piet-wgsl/shader/backdrop.twgsl:39]: Ok this is an interesting thing to expand. In the basic case, you want the workgroup size to be the width in pixels rounded up to the nearest power of 2. If that exceeds the maximum workgroup size (guaranteed to be at least 256, which is a dimension of 4k, so actually unlikely to be exceeded) then we can use N_SEQ > 1 (number of work items per thread, as in other prefix sum stages). Possibly N_SEQ > 1 would be a performance optimization in some cases even when we're not hitting the limit, but I wouldn't worry about it.

Getting that plumbed through is another question. According to the WGSL spec, we can use a pipeline override for workgroup, and I think that's the right thing. That's not supported by wgpu 0.14 though, so we should template in the meantime, with a TODO to switch to pipelines. (Using the pipeline approach would likely avoid recompilation to shader IR in the Metal case, where workgroup size can be late-bound on dispatch). That also interacts with precompiled shaders, but that should be discussed there I think.

File: piet-wgsl/shader/backdrop.twgsl [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+//
+// Also licensed under MIT license, at your choice.
+
+// Note: this is the non-atomic version
+struct Tile {
+    backdrop: i32,
+    segments: u32,
+}
+
+struct Config {
```
[the remark this answers is not in the document]
Comment (raphlinus) [reply piet-wgsl/shader/backdrop.twgsl:23]: Just simple oversight.

File: Cargo.toml [hunk 5]
```diff
@@ -8,5 +8,9 @@ members = [
     "piet-gpu-hal",
     "piet-gpu-types",
     "piet-scene",
+    "piet-wgsl",
     "tests",
 ]
+
+[patch.crates-io]
+naga = { path = "../../dl/naga" }
```
[the remark this answers is not in the document]
Comment (raphlinus) [reply Cargo.toml:16]: Yup, I didn't mean to commit that, thanks.

**Comment by raphlinus** [comment]
Thanks for the review, it's much appreciated!

Regarding templating, I'd be happy to delegate that to you. We chose handlebars after some discussion, and learning that @cwfitzgerald uses it for rend3 (I believe). This seems to be a concern across a wide range of the ecosystem, and I'm worried that the standards people will standardize on WebGPU/WGSL while leaving the layer immediately above to anarchy, following the traditional pattern. I've definitely been feeling the primitive IDE / highlighting / error reporting state and think it's important to improve that.

At least for right *now* I'm reluctant to go too deep into compile-time logic. Basically I'm comfortable with that handling includes, conditional computation, and basically permutation, but I am worried if it goes too far into traits, it will become an ad-hoc programming language that raises the cognitive load. Not that we shouldn't experiment with it (we should!) but that I'm resisting the temptation within piet-gpu. This might be a somewhat academic distinction, because we may well end up needing a powerful language to handle all the permutations correctly by the time we're done.

Also note that to some extent you do have the power of all of Rust in handlebars, as you can register arbitrarily powerful helper functions. Again, I'm resisting the temptation, as that makes things much harder to analyze without having deep tool support.

So basically, templating is a problem that needs solving well, and handlebars is a reasonable stopgap until then.

File: piet-wgsl/shader/backdrop.twgsl (hunk 1 above)
[the remark this answers is not in the document]
Comment (raphlinus) [reply piet-wgsl/shader/backdrop.twgsl:18]: Heh, wrong language though. I'll address it by adding a line to .gitattributes.

**Comment by raphlinus** [comment]
Going to merge this as-is. It's very much an experimental proof of concept, and there are no complex cross-dependencies, so we can consider the subdirectory to be a playground for now. I'm also actively working on porting more of the shaders.
